### PR TITLE
Deprecate telemetry

### PR DESCRIPTION
**Pull request by ricardoccpaiva**
As `Tesla` already has the concept of [Middlewares](https://github.com/teamon/tesla#middleware) there's no need to rely on [Elixir Telemetry](https://hex.pm/packages/telemetry).

### CODE REVIEW

**Comment by ricardoccpaiva** [comment]
The diff is not the best because I ended up moving some files.

**Comment by tsloughter** [comment]
Looks good to me :+1: and agree on just waiting on this before submitting to contrib.

The only possible place you might need telemetry is once OpenTelemetry Metrics are available. I don't know for sure, I'd need to know what all data tesla puts in the telemetry event, but if it has measurements of like "time getting connection from pool" that aren't available any other way then eventually will need telemetry.

For now I think this makes the most sense. And even if the additional (don't necessarily need Telsa to tell us how long the request took since we sort of get that from the Span) metrics are tracked but only reported through a telemetry event it might make sense to propose a patch to Tesla to make them a part of the result as well.

**Comment by dvic** [comment]
I think this makes sense 👍

**Comment by bryannaegele** [comment]
If we're going to do a major version then I'd also like to figure out if we need a hard dependency to get path params to work correctly and get the span name correct. https://hexdocs.pm/tesla/Tesla.Middleware.PathParams.html

That bit is pretty high on my list of things for work. It appears we can determine it's being used and can update the name accordingly by checking the opts. https://github.com/teamon/tesla/blob/v1.4.4/lib/tesla/middleware/path_params.ex#L31

**Comment by ricardoccpaiva** [comment]
> If we're going to do a major version then I'd also like to figure out if we need a hard dependency to get path params to work correctly and get the span name correct. https://hexdocs.pm/tesla/Tesla.Middleware.PathParams.html
> 
> That bit is pretty high on my list of things for work. It appears we can determine it's being used and can update the name accordingly by checking the opts. https://github.com/teamon/tesla/blob/v1.4.4/lib/tesla/middleware/path_params.ex#L31

@bryannaegele regarding your comments I think we can narrow it down to these 3 options and I'll write them down even if they don't make that much sense... 
Just to recap, our goal is to have for example `/users/:id` as the span name when `Tesla.Middleware.PathParams` it's being used and something else other than `/users/3` which is too specific when `Tesla.Middleware.PathParams` is **not** being used.

1. Add a key to the [middleware options parameter ](https://hexdocs.pm/tesla/Tesla.Middleware.html#c:call/3) to indicate if the PathParams Middleware is being used or not. Don't think we have a way to detect if a specific Middleware is being used or not.
2. Rely on order and always set `Tesla.Middleware.PathParams` **after** `Tesla.Middleware.OpenTelemetry`. With this approach we can look at the `next` argument of the `Tesla.Middleware.call/3` callback and know whether is being used or not. As we all know, relying on order is not exactly the best solution....
3. Modify `Tesla.Middleware.PathParams` middleware and probably `Tesla.Env` to include some sort of explicit field that stores the `parameterized` version of the path. Not sure how appropriate this change it is since it's too specific to solve a single problem

Speaking of bumping this library to a major version... maybe we should keep both options and have a deprecation window instead, what do you think?

cc @dvic @tsloughter

**Comment by dvic** [comment]
And what about making `Tesla.Middleware.OpenTelemetry` configurable to also implement the behaviour of `Tesla.Middleware.PathParams`? It doesn't look like too much trouble: https://github.com/teamon/tesla/blob/v1.4.4/lib/tesla/middleware/path_params.ex#L1

Injecting headers would then also become an option (with a default value of course). The only downside of this is that users would have to remove the `Tesla.Middleware.PathParams` middleware or make sure `Tesla.Middleware.OpenTelemetry` is run first.

**Comment by ricardoccpaiva** [comment]
or maybe this can help is 👇 
https://hexdocs.pm/tesla/Tesla.Middleware.KeepRequest.html

**Comment by ricardoccpaiva** [comment]
> And what about making `Tesla.Middleware.OpenTelemetry` configurable to also implement the behaviour of `Tesla.Middleware.PathParams`? It doesn't look like too much trouble: https://github.com/teamon/tesla/blob/v1.4.4/lib/tesla/middleware/path_params.ex#L1
> 
> Injecting headers would then also become an option (with a default value of course). The only downside of this is that users would have to remove the `Tesla.Middleware.PathParams` middleware or make sure `Tesla.Middleware.OpenTelemetry` is run first.

not sure if I got your suggestion, what behaviour are you exactly talking about?

**Comment by ricardoccpaiva** [comment]
> @ricardoccpaiva I think we can infer it from that already existing key I linked to being present which would also tell us the middlewares are in the correct order.

Yes we can, but it still requires a specific order when setting up the multiple middlewares. But it's ok for me, it's a tradeoff for having better span names.

**Comment by andrewhr** [comment]
> The only possible place you might need telemetry is once OpenTelemetry Metrics are available. I don't know for sure, I'd need to know what all data tesla puts in the telemetry event, but if it has measurements of like "time getting connection from pool" that aren't available any other way then eventually will need telemetry.

@tsloughter for the record, I think this use case is probably delegated to HTTP backends. Example, this information would be extracted from [Finch telemetry events](https://github.com/sneako/finch/blob/f8ecadd1f62d855a3e925805203df174a49955d6/lib/finch/telemetry.ex#L9-L34).

**Comment by bryannaegele** [comment]
> Yes we can, but it still requires a specific order when setting up the multiple middlewares. But it's ok for me, it's a tradeoff for having better span names.

Yes, I think this is ok. We should call it out in the documentation.

@andrewhr I don't think for Tesla middleware we necessarily want to offload to another lib for the various backends which would require the user to add yet another lib. There are also backends which do not have telemetry, e.g. hackney, so relying on Tesla telemetry for metrics if it we need to makes the most sense. We won't want to say you picked hackney, so no metrics for you.

**Comment by andrewhr** [comment]
@bryannaegele sorry, I've expressed the idea poorly or incomplete. I think "request-specific" metrics/traces done in this library is the way to go as that gets a huge coverage no matter what backend you choose. So in your example, Tesla + Hackney still gets the metrics and spans about the client-side of an HTTP request.

My earlier comment was about the cases of implementation specific metrics. Like in the Finch case, how it deals with pools is specific to its own impl (and Tesla doesn't know about it). Those specific telemetry events could be tracked in `opentelemetry_finch` for example. Hackney would not have anything similar, because doesn't provide this "meta" telemetry. 

In other words, it's a complement for HTTP spans created by Tesla instrumentation, not a replacement.

Hope that makes a bit more sense.

**Comment by ricardoccpaiva** [comment]
> My earlier comment was about the cases of implementation specific metrics. Like in the Finch case, how it deals with pools is specific to its own impl (and Tesla doesn't know about it). Those specific telemetry events could be tracked in `opentelemetry_finch` for example. Hackney would not have anything similar, because doesn't provide this "meta" telemetry.
> 
> In other words, it's a complement for HTTP spans created by Tesla instrumentation, not a replacement.
> 
> Hope that makes a bit more sense.

This makes total sense to me 👍

**Comment by ricardoccpaiva** [comment]
> Yes, I think this is ok. We should call it out in the documentation.

Yes!

**Comment by ricardoccpaiva** [comment]
> Speaking of bumping this library to a major version... maybe we should keep both options and have a deprecation window instead, what do you think?

what are your opinions on this?

**Comment by ricardoccpaiva** [comment]
with the changes in this PR client span names will look good :innocent: 
image

**Review by sashaafm — APPROVED** [review]
LGTM

**Comment by tsloughter** [comment]
Not sure its worth keeping both options unless moving to the new version would require changes to how someone includes the middleware. Which, I don't think it does?

**Comment by ricardoccpaiva** [comment]
> Not sure its worth keeping both options unless moving to the new version would require changes to how someone includes the middleware. Which, I don't think it does?

The only thing to keep in mind when using the middleware is the order of the `PathParams` + `OpenTelemetry`.

**Comment by bryannaegele** [comment]
Looks good. Let's just get some documentation.

**Review by dvic — APPROVED** [review]

**Review by bryannaegele — APPROVED** [review]
